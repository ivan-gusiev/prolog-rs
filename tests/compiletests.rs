extern crate insta;
extern crate parameterized;
extern crate prolog_rs;

#[cfg(test)]
mod compiletests {
    use insta::assert_display_snapshot;
    use parameterized::parameterized;
    use prolog_rs::{
        compile::{compile_program, compile_query},
        lang::parse_term,
        symbol::SymbolTable,
        util::{case, writeout_sym},
    };

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "a"
    })]
    fn test_query_compile(input: &str) {
        let mut symbol_table = SymbolTable::new();
        let query = parse_term(input, &mut symbol_table).unwrap();
        let instructions = compile_query(query);

        assert_display_snapshot!(case(input, writeout_sym(&instructions, &symbol_table)));
    }

    #[parameterized(input = {
        "p(Z, h(Z,W), f(W))",
        "D",
        "f(X, g(X,a))",
        "f(b, Y)",
        "p(f(X), h(Y, f(a)), Y)",
        "a"
    })]
    fn test_program_compile(input: &str) {
        let mut symbol_table = SymbolTable::new();
        let query = parse_term(input, &mut symbol_table).unwrap();
        let instructions = compile_program(query);

        assert_display_snapshot!(case(input, writeout_sym(&instructions, &symbol_table)));
    }
}
