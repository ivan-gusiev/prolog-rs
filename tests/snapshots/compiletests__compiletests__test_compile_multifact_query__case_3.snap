---
source: tests/compiletests.rs
expression: "case(input, collapse(result))"
---
?- p(a, X, Y), f(b, f(X)).
-----
000	allocate 2
001	put_structure a/0, X1
002	put_variable Y1, X2	// Y1=X
003	put_variable Y2, X3	// Y2=Y
004	call @0
005	put_structure b/0, X1
006	put_structure f/1, X2
007	set_value Y1	// Y1=X
008	call @10
009	publish
010	deallocate

